### PR TITLE
Add simple missing inferShapes

### PR DESCRIPTION
**Pull request by philass**
I'm quite invested in closing the gap on missing `inferShapes`.

Some of the `inferShapes` are not currently implmented because `Transform/ONNX/decompose.td` converts these ops into an other op or a combination of other ops. This is a problem when `onnx-mlir` is used as a dependency for `onnx` mlir ingestion, as it creates a pass ordering dependency. The `ONNX Decomposition` must be run before the `shapeInference` pass. Thats why I believe its worth adding their shape-inference implementations. 

This PR adds support for some very simple missing `inferShapes`. 

These shape inferences fall into the falling categories:
- result has the same shape as its input
- result shape can be derived from binary broadcasting (BitShift op)
- An already existing Op has the exact same `inferShape` implementation

I haven't added shape inference `lit` tests because I haven't added any logic. As in I use/copy code thats already been tested by other Ops' shape inference tests. If you guys need something from me on this front, please let me know and I'll be happy to add it!

----

On a side note I plan on adding many more `shapeInferences` but I wanted to start with the simple ones, to start the process. Hopefully we can  get them all in the next month or two :)

### CODE REVIEW

**Comment by AlexandreEichenberger** [comment]
@chentong319 I think you left a "LGTM" comment but not officially approve it. Can you do it?

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 1]
```diff
@@ -1685,6 +1685,76 @@ LogicalResult ONNXReduceProdOp::inferShapes(
...
+
+LogicalResult ONNXReduceSumSquareOp::inferShapes(
+    std::function<void(mlir::Region &)> doShapeInference) {
+  if (!getOperand().getType().isa<RankedTensorType>())
+    return success();
+
+  auto operandTy = getOperand().getType().cast<RankedTensorType>();
+  getResult().setType(getReductionOutputType(operandTy, axes(), keepdims()));
+  return success();
+}
```
Comment (etiotto) [inline src/Dialect/ONNX/ONNXOps.cpp:1756]: See comment at line 1692

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 2]
```diff
@@ -4093,12 +4164,14 @@ LogicalResult ONNXLogSoftmaxOp::verify() {
 
 LogicalResult ONNXLogSoftmaxOp::inferShapes(
     std::function<void(mlir::Region &)> doShapeInference) {
-  return emitError(NOT_IMPLEMENTED_MESSAGE);
+  getResult().setType(getOperand().getType());
```
[the remark this answers is not in the document]
Comment (philass) [reply src/Dialect/ONNX/ONNXOps.cpp:4167]: I copied this code pattern from `ExpOp`. It is used in 10+ ops

Its used in
- `Exp`
- `Atan`
- `Sin`
- `Sinh`
and many more ops. If you think this is the way to go, I can add this check to all the many other ops that use this pattern.

I thought that   `getResult().setType(getOperand().getType());` was kind of a catch all. As in if the type of `getOperand()` didn't have shape or rank, we would just be setting the result type to the same unranked and or unshaped type.

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 3]
```diff
@@ -1685,6 +1685,76 @@ LogicalResult ONNXReduceProdOp::inferShapes(
       *this);
 }
 
+//===----------------------------------------------------------------------===//
+// ReduceL1
+//===----------------------------------------------------------------------===//
+
+LogicalResult ONNXReduceL1Op::inferShapes(
```
[the remark this answers is not in the document]
Comment (philass) [reply src/Dialect/ONNX/ONNXOps.cpp:1692]: Nice addition!

I've updated this PR to use the `inferShapeForReductionOps` function

[the remark this answers is not in the document]
Comment (etiotto) [reply src/Dialect/ONNX/ONNXOps.cpp:1692]: Thank @philass !

**Review by etiotto — APPROVED** [review]
LGTM - Thanks for your contribution @philass

**Comment by philass** [comment]
@AlexandreEichenberger Thank you! I appreciate that.

Would you mind rerunning the CI, seems that I don't have the permissions for that.

I am done with the changes, if you could merge this for me once it passes CI, that would be awesome.

**Comment by philass** [comment]
@AlexandreEichenberger, @etiotto

Could you guys merge this into main for me?

**Pull request merged**